# "Data Cleaning: A Crucial Preparatory Step in Exploratory Data Analysis"

# Intruduction

In this segment of our Exploratory Data Analysis (EDA), we are embarking on a process that involves cleaning and performing feature engineering on our aggregated dataset derived from separate CSV files. Upon completion, we will generate a new, cleaned dataset to facilitate the continuation of our EDA.Initially, we will focus on identifying and handling missing and duplicated values within our dataset. Subsequently, we will proceed to engineer a new feature named "Sales Revenue", computed by multiplying the item quantity by the item price. Lastly, we aim to transform categorical variables, specifically 'State' and 'Payment Type', into numerical equivalents for better data manipulation and analysis.

In the first step, we import the necessary libraries.

In [1]:
# import the relevant libraries
import numpy as np
import pandas as pd
from pandas import DataFrame
from sklearn.preprocessing import LabelEncoder

## Import the Dataset

Next, we will proceed to load our newly created dataset.

In [2]:
# Load the data 
df = pd.read_csv('Olist_CombinedData.csv')
df. head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  130898c0987d1801452a8ed92a670612              1   
2  532ed5e14e24ae1f0d735b91524b98b9              1   
3  6f8c31653edb8c83e1a739408b5ff750              1   
4  7d19f4ef4d04461989632411b7e588b9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
2  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
3  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
4  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   

  shipping_limit_date  price  freight_value  payment_sequential payment_type  \
0     2017-09-19 9:45   58.9          13.29                   1  credit_card   
1     2017-07-05 2:44   55.9          17.96                   1       boleto   
2    2018-05-23 10:56   64.9          18.33                   1  credit_card   
3    2017-08-07 18:55   58.9          16.17                   1  credit_card   
4    2017-08-16 22:05   58.9          13.29                   1  credit_card   

   payment_installments  ...  product_name_lenght product_description_lenght  \
0                     2  ...                 58.0                      598.0   
1                     1  ...                 58.0                      598.0   
2                     2  ...                 58.0                      598.0   
3                     3  ...                 58.0                      598.0   
4                     4  ...                 58.0                      598.0   

  product_photos_qty product_weight_g product_length_cm product_height_cm  \
0                4.0            650.0              28.0               9.0   
1                4.0            650.0              28.0               9.0   
2                4.0            650.0              28.0               9.0   
3                4.0            650.0              28.0               9.0   
4                4.0            650.0              28.0               9.0   

  product_width_cm seller_zip_code_prefix    seller_city  seller_state  
0             14.0                  27277  volta redonda            SP  
1             14.0                  27277  volta redonda            SP  
2             14.0                  27277  volta redonda            SP  
3             14.0                  27277  volta redonda            SP  
4             14.0                  27277  volta redonda            SP  

[5 rows x 39 columns]

Let's first take a look at the information about our dataset.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 117329 entries, 0 to 117328
Data columns (total 39 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   order_id                       117329 non-null  object 
 1   order_item_id                  117329 non-null  int64  
 2   product_id                     117329 non-null  object 
 3   seller_id                      117329 non-null  object 
 4   shipping_limit_date            117329 non-null  object 
 5   price                          117329 non-null  float64
 6   freight_value                  117329 non-null  float64
 7   payment_sequential             117329 non-null  int64  
 8   payment_type                   117329 non-null  object 
 9   payment_installments           117329 non-null  int64  
 10  payment_value                  117329 non-null  float64
 11  customer_id                    117329 non-null  object 
 12  order_status                  

The dataset contains 117,329 entries, indexed from 0 to 117,328. It encompasses a total of 39 columns, each of which holds data of various types.

Refine the dataset by removing the columns that are not pertinent to our analysis.

In [4]:
# Narrowing the dataset
df=df.drop (['shipping_limit_date','product_name_lenght',
          'product_description_lenght','product_photos_qty','order_status','order_approved_at','order_delivered_carrier_date', 'order_delivered_customer_date',
          'order_estimated_delivery_date','review_comment_title','review_comment_message',
          'review_creation_date','review_answer_timestamp','customer_unique_id(barcode)','seller_city','review_id','seller_zip_code_prefix','city','seller_state'], axis=1)
df.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  130898c0987d1801452a8ed92a670612              1   
2  532ed5e14e24ae1f0d735b91524b98b9              1   
3  6f8c31653edb8c83e1a739408b5ff750              1   
4  7d19f4ef4d04461989632411b7e588b9              1   

                         product_id                         seller_id  price  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   58.9   
1  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   55.9   
2  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   64.9   
3  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   58.9   
4  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   58.9   

   freight_value  payment_sequential payment_type  payment_installments  \
0          13.29                   1  credit_card                     2   
1          17.96                   1       boleto                     1   
2          18.33                   1  credit_card                     2   
3          16.17                   1  credit_card                     3   
4          13.29                   1  credit_card                     4   

   payment_value                       customer_id order_purchase_timestamp  \
0          72.19  3ce436f183e68e07877b285a838db11a          2017-09-13 8:59   
1          73.86  e6eecc5a77de221464d1c4eaff0a9b64         2017-06-28 11:52   
2          83.23  4ef55bf80f711b372afebcb7c715344a         2018-05-18 10:25   
3          75.07  30407a72ad8b3f4df4d15369126b20c9         2017-08-01 18:38   
4          72.19  91a792fef70ecd8cc69d3c7feb3d12da         2017-08-10 21:48   

   review_score  postal_code state product_category_name  product_weight_g  \
0             5        28013    RJ            cool_stuff             650.0   
1             5        75800    GO            cool_stuff             650.0   
2             4        30720    MG            cool_stuff             650.0   
3             5        83070    PR            cool_stuff             650.0   
4             5        36400    MG            cool_stuff             650.0   

   product_length_cm  product_height_cm  product_width_cm  
0               28.0                9.0              14.0  
1               28.0                9.0              14.0  
2               28.0                9.0              14.0  
3               28.0                9.0              14.0  
4               28.0                9.0              14.0

## Duplicate and Null Value Handeling

Next, we will check for any duplicate rows in our dataset. If any are found, we will proceed to eliminate them, retaining only one instance.

In [5]:
# Count the number of duplicated rows
num_duplicates = df.duplicated().sum()

# Display the total number of duplicated rows
print("Total number of duplicated rows:", num_duplicates)

# Remove duplicate rows
df = df.drop_duplicates()

# Count the number of duplicated rows
num_duplicates = df.duplicated().sum()

# Display the total number of duplicated rows
print("Total number of duplicated rows:", num_duplicates)

Total number of duplicated rows: 419
Total number of duplicated rows: 0


We initially found 271 duplicate rows in our dataset. We then proceeded to remove these duplicates using the drop_duplicates() function. After retesting with the df.duplicated().sum() command, it confirmed that we now have no duplicate rows in our dataset.

Now, we will proceed to check for any null values within the dataset.

In [6]:
df.isnull().sum().sum()

1769

In [7]:
df.isna().sum()/df.shape[0]*100

order_id                    0.000000
order_item_id               0.000000
product_id                  0.000000
seller_id                   0.000000
price                       0.000000
freight_value               0.000000
payment_sequential          0.000000
payment_type                0.000000
payment_installments        0.000000
payment_value               0.000000
customer_id                 0.000000
order_purchase_timestamp    0.000000
review_score                0.000000
postal_code                 0.000000
state                       0.000000
product_category_name       1.444701
product_weight_g            0.017107
product_length_cm           0.017107
product_height_cm           0.017107
product_width_cm            0.017107
dtype: float64

We have detected a total of 1769 null values in our dataset. The columns 'product_category_name', 'product_weight_g', 'product_length_cm', 'product_height_cm', and 'product_width_cm' have relatively low percentages of null values, ranging from 0.017% to 1.44%. Given their low impact, we have decided to drop these null values.

In [8]:
# Drop rows in 'df' with NaN in specific columns.
df =df.dropna(subset=['product_category_name','product_weight_g','product_length_cm','product_height_cm','product_width_cm'])

Upon rechecking for null values in our dataset, we can now confirm that there are no more null entries remaining. This successfully concludes our data cleaning process in terms of handling missing values.

In [9]:
df.isna().sum()/df.shape[0]*100

order_id                    0.0
order_item_id               0.0
product_id                  0.0
seller_id                   0.0
price                       0.0
freight_value               0.0
payment_sequential          0.0
payment_type                0.0
payment_installments        0.0
payment_value               0.0
customer_id                 0.0
order_purchase_timestamp    0.0
review_score                0.0
postal_code                 0.0
state                       0.0
product_category_name       0.0
product_weight_g            0.0
product_length_cm           0.0
product_height_cm           0.0
product_width_cm            0.0
dtype: float64

Given that our 'product_category_name' entries are in a foreign language, and we possess a translated list of these categories in English, we will map the foreign language categories to their English counterparts. We will then use the 'replace' function to substitute the foreign language category names with the corresponding English ones in our dataset.

In [10]:
# Create a dictionary 'name_mapping' to replace the product_category_names. 
name_mapping = {
    'beleza_saude': 'health_beauty',
    'informatica_acessorios': 'computers_accessories',
    'automotivo': 'auto',
    'cama_mesa_banho': 'bed_bath_table',
    'moveis_decoracao': 'furniture_decor',
    'esporte_lazer': 'sports_leisure',
    'perfumaria': 'perfumery',
    'utilidades_domesticas': 'housewares',
    'telefonia': 'telephony',
    'relogios_presentes': 'watches_gifts',
    'alimentos_bebidas': 'food_drink',
    'bebes': 'baby',
    'papelaria': 'stationery',
    'tablets_impressao_imagem': 'tablets_printing_image',
    'brinquedos': 'toys',
    'telefonia_fixa': 'fixed_telephony',
    'ferramentas_jardim': 'garden_tools',
    'fashion_bolsas_e_acessorios': 'fashion_bags_accessories',
    'eletroportateis': 'small_appliances',
    'consoles_games': 'consoles_games',
    'audio': 'audio',
    'fashion_calcados': 'fashion_shoes',
    'cool_stuff': 'cool_stuff',
    'malas_acessorios': 'luggage_accessories',
    'climatizacao': 'air_conditioning',
    'construcao_ferramentas_construcao': 'construction_tools_construction',
    'moveis_cozinha_area_de_servico_jantar_e_jardim': 'kitchen_dining_laundry_garden_furniture',
    'construcao_ferramentas_jardim': 'costruction_tools_garden',
    'fashion_roupa_masculina': 'fashion_male_clothing',
    'pet_shop': 'pet_shop',
    'moveis_escritorio': 'office_furniture',
    'market_place': 'market_place',
    'eletronicos': 'electronics',
    'eletrodomesticos': 'home_appliances',
    'artigos_de_festas': 'party_supplies',
    'casa_conforto': 'home_confort',
    'construcao_ferramentas_ferramentas': 'costruction_tools_tools',
    'agro_industria_e_comercio': 'agro_industry_and_commerce',
    'moveis_colchao_e_estofado': 'furniture_mattress_and_upholstery',
    'livros_tecnicos': 'books_technical',
    'casa_construcao': 'home_construction',
    'instrumentos_musicais': 'musical_instruments',
    'moveis_sala': 'furniture_living_room',
    'construcao_ferramentas_iluminacao': 'construction_tools_lights',
    'industria_comercio_e_negocios': 'industry_commerce_and_business',
    'alimentos': 'food',
    'artes': 'art',
    'moveis_quarto': 'furniture_bedroom',
    'livros_interesse_geral': 'books_general_interest',
    'construcao_ferramentas_seguranca': 'construction_tools_safety',
    'fashion_underwear_e_moda_praia': 'fashion_underwear_beach',
    'fashion_esporte': 'fashion_sport',
    'sinalizacao_e_seguranca': 'signaling_and_security',
    'pcs': 'computers',
    'artigos_de_natal': 'christmas_supplies',
    'fashion_roupa_feminina': 'fashio_female_clothing',
    'eletrodomesticos_2': 'home_appliances_2',
    'livros_importados': 'books_imported',
    'bebidas': 'drinks',
    'cine_foto': 'cine_photo',
    'la_cuisine': 'la_cuisine',
    'musica': 'music',
    'casa_conforto_2': 'home_comfort_2',
    'portateis_casa_forno_e_cafe': 'small_appliances_home_oven_and_coffee',
    'cds_dvds_musicais': 'cds_dvds_musicals',
    'dvds_blu_ray': 'dvds_blu_ray',
    'flores': 'flowers',
    'artes_e_artesanato': 'arts_and_craftmanship',
    'fraldas_higiene': 'diapers_and_hygiene',
    'fashion_roupa_infanto_juvenil': 'fashion_childrens_clothes',
    'seguros_e_servicos': 'security_and_services',
}
df['product_category_name'] = df['product_category_name'].replace(name_mapping)

## Feature Engineering

In this step, we aim to transform our 'payment_type' and 'state' variables into numerical values. For this task, we will employ the LabelEncoder class from the sklearn.preprocessing module. The LabelEncoder will assign each unique category in these columns a distinct numerical value, thereby allowing for a more streamlined analysis.

In [11]:
# Initialize a LabelEncoder
le = LabelEncoder()

# Convert the 'payment_type' column into numeric
df['payment_type'] = le.fit_transform(df['payment_type'])

# Convert the 'state' column into numeric
df['state'] = le.fit_transform(df['state'])

In our next move, we will perform feature engineering on our data frame. We will create a new feature named 'sales_revenue'. This new feature will be calculated using two existing features in our dataset: 'order_item_id' and 'price'.
By multiplying these two features together for each row in our dataframe, we can calculate the total sales revenue per order. This will be stored in our new 'sales_revenue' column.

In [12]:
# Create new feature 'sales_revenue' of 'order_item_id' and 'price'.
df['sales_revenue'] = df['order_item_id'] * df['price']

Lastly, we are enhancing the intuitiveness of our dataframe by renaming certain columns. Specifically, we are changing 'order_item_id' to 'item' and 'price' to 'item_price'. These adjustments better represent the information contained within these columns: 'item' indicates the quantity of each item in an order, while 'item_price' represents the price per item. These modifications will improve clarity and ease future analysis or model building tasks. Finally, we will save our cleaned and refined dataframe as a new CSV file in our directory.

In [13]:
# Rename multiple columns
df.rename(columns={'order_item_id': 'item', 'price': 'item_price'}, inplace=True)

Lastly, the processed data frame is saved and exported as a CSV file to the designated location.

In [14]:
# Save the merged and cleaned data to a new file for use it for further analysis
df.to_csv('Olist_clean_data.csv', index=False)

# Conclusion

In conclusion, this notebook began with importing the dataset, followed by a data cleaning process where unnecessary columns were removed. We also handled duplicate and null values, eliminating them to refine our dataset. Furthermore, we performed feature engineering to enhance the quality of our dataset, thus preparing it for more accurate and efficient modeling. Finally, the cleaned and enriched dataset was saved as a new CSV file in our directory.